// ngx-bar-rating/esm2022/lib/custom-rating.mjs
import { Directive } from '@angular/core';
import * as i0 from "@angular/core";
export class ActiveRating {
    constructor(template) {
        this.template = template;
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.0.2", ngImport: i0, type: ActiveRating, deps: [{ token: i0.TemplateRef }], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "16.0.2", type: ActiveRating, selector: "[ratingActive]", ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.0.2", ngImport: i0, type: ActiveRating, decorators: [{
            type: Directive,
            args: [{
                    selector: '[ratingActive]'
                }]
        }], ctorParameters: function () { return [{ type: i0.TemplateRef }]; } });
export class InactiveRating {
    constructor(template) {
        this.template = template;
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.0.2", ngImport: i0, type: InactiveRating, deps: [{ token: i0.TemplateRef }], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "16.0.2", type: InactiveRating, selector: "[ratingInactive]", ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.0.2", ngImport: i0, type: InactiveRating, decorators: [{
            type: Directive,
            args: [{
                    selector: '[ratingInactive]'
                }]
        }], ctorParameters: function () { return [{ type: i0.TemplateRef }]; } });
export class FractionRating {
    constructor(template) {
        this.template = template;
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.0.2", ngImport: i0, type: FractionRating, deps: [{ token: i0.TemplateRef }], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "16.0.2", type: FractionRating, selector: "[ratingFraction]", ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.0.2", ngImport: i0, type: FractionRating, decorators: [{
            type: Directive,
            args: [{
                    selector: '[ratingFraction]'
                }]
        }], ctorParameters: function () { return [{ type: i0.TemplateRef }]; } });
